Add support for JWT authentication

BaurzhanSakhariev opened the pull request:
Closes https://github.com/crate/crate/issues/14238


Each commit passes tests (rebased with `git rebase -i --exec "./mvnw test" origin/master`)

First commit is combination of https://github.com/crate/crate/pull/15515 and part of https://github.com/crate/crate/pull/15562 (another part is exposure of props in sys.users table)

BaurzhanSakhariev commented:
unrelated LR failure 
https://jenkins.crate.io/job/CrateDB/job/crate_on_pr/13302/testReport/junit/io.crate.integrationtests/MetadataTrackerITest/test_schema_changes_of_subscribed_table_is_replicated/

retest this please

matriv commented on server/src/main/java/io/crate/exceptions/RoleAlreadyExistsException.java:
```diff
@@ -23,9 +23,25 @@
+        super(message);
+    }
+
+    public static RoleAlreadyExistsException of(String roleName, @Nullable JwtProperties jwtProperties) {
```
minor: I would pass a boolean flag, to denote the kind of the exception and not a jwtProperties arg.

matriv commented on server/src/test/java/io/crate/role/UserActionsTest.java:
```diff
@@ -58,39 +71,60 @@ public void testSecureHashIsGeneratedFromPasswordProperty() throws Exception {
+    public void test_invalid_jwt_property() throws Exception {
+        // Missing jwt property
+        GenericProperties<Symbol> properties = new GenericProperties<>(Map.of("jwt", Literal.of(Map.of("iss", "dummy.org"))));
+        Map<String, Object> parsedProperties = CreateRolePlan.parse(properties, txnCtx, NODE_CTX, Row.EMPTY, SubQueryResults.EMPTY);
```
same here, consider adding this test to CreateRolePlanTest?

matriv commented on server/src/test/java/io/crate/role/UserActionsTest.java:
```diff
@@ -58,39 +71,60 @@ public void testSecureHashIsGeneratedFromPasswordProperty() throws Exception {
     public void testInvalidPasswordProperty() throws Exception {
         GenericProperties<Symbol> properties = new GenericProperties<>(Map.of("invalid", Literal.of("password")));
-        assertThatThrownBy(() -> UserActions.generateSecureHash(properties, Row.EMPTY, txnCtx, NODE_CTX))
+        assertThatThrownBy(() -> CreateRolePlan.parse(properties, txnCtx, NODE_CTX, Row.EMPTY, SubQueryResults.EMPTY))
```
why user the Plan here since we're testing `UserActions`?

matriv commented on server/src/test/java/io/crate/integrationtests/RoleManagementIntegrationTest.java:
```diff
@@ -277,4 +277,14 @@ public void testDropSuperUserThrowsException() {
     }
+
+    @Test
+    public void test_create_user_jwt_properties_must_be_unique() {
```
Can we have a test with different jwt properties, but with the same username (so user itself already exists, and not clash with existing jwt properties).

BaurzhanSakhariev replied:
I pushed only unit test. Tested locally that such scenario shows proper message
`Role  user1 or another role with the same combination of jwt properties already exists`

matriv commented on server/src/main/java/io/crate/role/RoleManager.java:
```diff
@@ -41,7 +41,11 @@ public interface RoleManager extends Roles {
      * @param roleName name of the role to create
```
can you please update the javadoc with the new params?

matriv commented on server/src/main/java/io/crate/protocols/http/Headers.java:
```diff
@@ -64,15 +64,30 @@ public static void setKeepAlive(HttpVersion httpVersion, FullHttpResponse resp)
         }
+        String[] splitHeader = authHeaderValue.split(" ");
+        assert splitHeader.length == 2 :
+            "Header must contain only authentication schema and base64 encoded value, separated by a whitespace";
```
```suggestion
            "Header must contain only authentication scheme and base64 encoded value, separated by a whitespace";
```

matriv replied:
in docs with name it  `method`, so maybe change it to `method` here as well?

BaurzhanSakhariev replied:
`method` in HBA is specific for our auth methods.
scheme is headers "key" name, either basic -> password or bearer -> jwt.
it's used in spec, https://datatracker.ietf.org/doc/html/rfc7235?ref=blog.teknkl.com#section-2.1. I adjusted header javadocs.

matriv commented on server/src/main/java/io/crate/protocols/http/Headers.java:
```diff
@@ -64,15 +64,30 @@ public static void setKeepAlive(HttpVersion httpVersion, FullHttpResponse resp)
+            case "basic" -> extractCredentialsFromHttpBasicAuthHeader(splitHeader[1]);
+            case "bearer" -> new Credentials(splitHeader[1]);
+            default ->
+                    throw new IllegalArgumentException("Only basic or bearer HTTP Authentication schemas are allowed.");
```
```suggestion
                    throw new IllegalArgumentException("Only basic or bearer HTTP Authentication schemes are allowed.");
```

matriv commented on server/src/main/java/io/crate/role/Role.java:
```diff
@@ -65,17 +69,23 @@ public static Properties fromXContent(XContentParser parser) throws IOException
-            this(in.readBoolean(), in.readOptionalWriteable(SecureHash::readFrom));
+            this(in.readBoolean(),
+                 in.readOptionalWriteable(SecureHash::readFrom),
+                 in.getVersion().onOrAfter(Version.V_5_7_0) ? in.readOptionalWriteable(JwtProperties::readFrom) : null
```
here we use `null`, but elsewhere (requests) we use empty maps, maybe we should be consistent?

BaurzhanSakhariev replied:
We use nullable record. I think there was a confusing fro the old implementation where Map was used, I removed it in https://github.com/crate/crate/pull/15640/commits/5bd489a2379ac72a9effbea9d881e1f5dc685268#diff-05ff3ea28f321d32053d4ba4043a7006962031aed40b321d33df54ba72620317L88

matriv replied:
:+1:

matriv commented on server/src/test/java/io/crate/auth/HttpAuthUpstreamHandlerTest.java:
```diff
@@ -134,7 +143,7 @@ public void testAuthorized() throws Exception {
     @Test
     public void testNotNoHbaConfig() throws Exception {
-        HttpAuthUpstreamHandler handler = new HttpAuthUpstreamHandler(Settings.EMPTY, authService);
+        HttpAuthUpstreamHandler handler = new HttpAuthUpstreamHandler(Settings.EMPTY, authService, new StubRoleManager());
```
consider extracting a var since this version used in many tests.

matriv commented on server/src/test/java/io/crate/auth/UserAuthenticationMethodTest.java:
```diff
@@ -70,33 +97,184 @@ public void testAlwaysOKAuthentication() throws Exception {
+
+    @Test
+    @SuppressWarnings("resource")
+    public void test_jwt_authentication_user_not_found_throws_an_error() throws Exception {
```
minor
```suggestion
    public void test_jwt_authentication_user_not_found_throws_error() throws Exception {
```

matriv commented on server/src/test/java/io/crate/auth/UserAuthenticationMethodTest.java:
```diff
@@ -70,33 +97,184 @@ public void testAlwaysOKAuthentication() throws Exception {
+
+    @Test
+    @SuppressWarnings("resource")
+    public void test_token_algo_and_jwk_algo_mistmatch_error_thrown() throws Exception {
```
for consistency:
```suggestion
    public void test_token_algo_and_jwk_algo_mistmatch_throws_error() throws Exception {
```

matriv commented on server/src/test/java/io/crate/auth/UserAuthenticationMethodTest.java:
```diff
@@ -70,33 +97,184 @@ public void testAlwaysOKAuthentication() throws Exception {
+
+    @Test
+    @SuppressWarnings("resource")
+    public void test_jwt_authentication_claim_mismatch_throws_an_error() throws Exception {
```
minor:
```suggestion
    public void test_jwt_authentication_claim_mismatch_throws_error() throws Exception {
```

matriv reviewed (commented):
thank you for all the work here @BaurzhanSakhariev 
Left a bunch of comments, but most of them are minor/cosmetic.

BaurzhanSakhariev commented on server/src/test/java/io/crate/auth/JwtAuthenticationIntegrationTest.java:
```diff
@@ -0,0 +1,213 @@
+        assertThat(bodyAsString).containsIgnoringWhitespaces("""
+                                                {
+                                                  "ok" : true,
+                                                  "status" : 200
```
yep, version, cluster name
```
{
  "ok" : true,
  "status" : 200,
  "name" : "node_sm0",
  "cluster_name" : "SUITE-TEST_WORKER_VM=[$1]-CLUSTER_SEED=[5481774869293243506]-HASH=[951BF04A409]-cluster",
  "version" : {
    "number" : "5.7.0",
    "build_hash" : "NA",
    "build_timestamp" : "NA",
    "build_snapshot" : true,
    "lucene_version" : "9.10.0"
  }
}
```

matriv commented on server/src/main/java/io/crate/planner/node/ddl/CreateRolePlan.java:
```diff
@@ -53,19 +68,49 @@ public StatementType type() {
         }
 
-        roleManager.createRole(createRole.roleName(), createRole.isUser(), newPassword)
+        // Map is always not null to simplify streaming.
```
also invalid comment?

BaurzhanSakhariev replied:
already removed in https://github.com/crate/crate/pull/15640/commits/5bd489a2379ac72a9effbea9d881e1f5dc685268#diff-05ff3ea28f321d32053d4ba4043a7006962031aed40b321d33df54ba72620317L88

matriv commented on server/src/test/java/io/crate/protocols/http/HeadersTest.java:
```diff
@@ -35,42 +35,52 @@ public class HeadersTest {
+
+    @Test
+    @SuppressWarnings("resource")
+    public void test_unsupported_schema_throws_an_error() {
```
```suggestion
    public void test_unsupported_schema_throws_error() {
```

BaurzhanSakhariev commented:
> @matriv spot an issue, will push a fixup. Thanks for catching!

Actually it was confusing comment form the old implementation, removed in https://github.com/crate/crate/pull/15640/commits/5bd489a2379ac72a9effbea9d881e1f5dc685268#diff-05ff3ea28f321d32053d4ba4043a7006962031aed40b321d33df54ba72620317L88
and also a potential confusion of using same exception for role exists/jwt combination exists

See https://github.com/crate/crate/pull/15640#discussion_r1508704694 and https://github.com/crate/crate/pull/15562#discussion_r1490967147

Question is whether we want to introduce dedicated exception for jwt clash and avoid OR in error, 
ie make it either "name clash" or "jwt clash" but not "name or jwt"

BaurzhanSakhariev commented:
`ALTER USER` doesn't modify jwt properties, will fix that

BaurzhanSakhariev commented:
Hi @matriv, I addressed comments and squashed changes and applied suggestions into corresponding commits. 

Since ALTER ROLE is not supported and looks like going to be a bigger change, I will create a separate PR against feature branch

Some of comments will be be addressed in that change, for example everything about `RoleAlreadyExistsException` gonna be outdated as I will introduce dedicated exception. 

We can ALTER USER and try to assign existing jwt props, `RoleAlreadyExistsException` even with improved message would be confusing as of course user exists in that case, no need for 'OR"
